subsys/llext/llext.c

```diff
@@ -0,0 +1,469 @@
+				    sym.st_name);
+			llext_read(ldr, name, sizeof(name));
+
+			LOG_DBG("relocation %d:%d info %x (type %d, sym %d) offset %d sym_name "
+				"%s sym_type %d sym_bind %d sym_ndx %d",
+				i, j, rel.r_info, ELF_R_TYPE(rel.r_info), ELF_R_SYM(rel.r_info),
+				rel.r_offset, name, ELF_ST_TYPE(sym.st_info),
+				ELF_ST_BIND(sym.st_info), sym.st_shndx);
+
+			uintptr_t link_addr, op_loc, op_code;
+
+			/* If symbol is undefined, then we need to look it up */
+			if (sym.st_shndx == SHN_UNDEF) {
+				link_addr = (uintptr_t)llext_find_sym(&SYMTAB, name);
+
+				if (link_addr == 0) {
+					LOG_ERR("Undefined symbol with no entry in "
+						"symbol table %s, offset %d, link section %d",
+						name, rel.r_offset, shdr.sh_link);
+					/* TODO cleanup and leave */
+					continue;
+				} else {
+					op_code = (uintptr_t)(loc + rel.r_offset);
+
+					LOG_INF("found symbol %s at 0x%lx, updating op code 0x%lx",
+						name, link_addr, op_code);
+				}
+			} else if (ELF_ST_TYPE(sym.st_info) == STT_SECTION) {
+				link_addr = (uintptr_t)ext->mem[ldr->sect_map[sym.st_shndx]];
+
+				LOG_INF("found section symbol %s addr 0x%lx", name, link_addr);
+			} else {
+				/* Nothing to relocate here */
+				continue;
+			}
+
+			op_loc = loc + rel.r_offset;
+
+			LOG_INF("relocating (linking) symbol %s type %d binding %d ndx %d offset "
+				"%d link section %d",
+				name, ELF_ST_TYPE(sym.st_info), ELF_ST_BIND(sym.st_info),
+				sym.st_shndx, rel.r_offset, shdr.sh_link);
+
+			LOG_INF("writing relocation symbol %s type %d sym %d at addr 0x%lx "
+				"addr 0x%lx",
+				name, ELF_R_TYPE(rel.r_info), ELF_R_SYM(rel.r_info),
+				op_loc, link_addr);
+
+			/* relocation */
+			arch_elf_relocate(&rel, op_loc, link_addr);
+		}
+	}
+
+	LOG_DBG("loaded module, .text at %p, .rodata at %p",
+		ext->mem[LLEXT_MEM_TEXT], ext->mem[LLEXT_MEM_RODATA]);
+	return 0;
+}
+
+STRUCT_SECTION_START_EXTERN(llext_symbol);
+
+int llext_load(struct llext_loader *ldr, const char *name, struct llext **ext)
+{
+	int ret = 0;
+	elf_ehdr_t ehdr;
+
+	if (!SYMTAB.sym_cnt) {
+		STRUCT_SECTION_COUNT(llext_symbol, &SYMTAB.sym_cnt);
+		SYMTAB.syms = STRUCT_SECTION_START(llext_symbol);
+	}
+
+	llext_seek(ldr, 0);
+	llext_read(ldr, &ehdr, sizeof(ehdr));
+
+	/* check whether this is an valid elf file */
+	if (memcmp(ehdr.e_ident, ELF_MAGIC, sizeof(ELF_MAGIC)) != 0) {
+		LOG_HEXDUMP_ERR(ehdr.e_ident, 16, "Invalid ELF, magic does not match");
+		return -EINVAL;
+	}
+
+	switch (ehdr.e_type) {
+	case ET_REL:
+	case ET_DYN:
+		LOG_DBG("Loading relocatable or shared elf");
+		*ext = k_heap_alloc(&llext_heap, sizeof(struct llext), K_NO_WAIT);
+
+		for (int i = 0; i < LLEXT_MEM_COUNT; i++) {
+			(*ext)->mem[i] = NULL;
+		}
+
+		if (ext == NULL) {
+			LOG_ERR("Not enough memory for extension metadata");
+			ret = -ENOMEM;
+		} else {
+			ldr->hdr = ehdr;
+			ret = llext_load_rel(ldr, *ext);
+		}
+		break;
+	default:
+		LOG_ERR("Unsupported elf file type %x", ehdr.e_type);
+		*ext = NULL;
```

Comment:
I think you need to set `ret` to a negative error here, otherwise line 427 below will dereference `NULL`